test/aleph/resource_leak_detector.clj

```diff
@@ -0,0 +1,210 @@
+  Since it adds considerable runtime overhead, its main purpose is to be used in tests to assert
+  intentional presence of leaks and to check for unintended leaks. The main API entry points for
+  this are `with-leak-collection`, `with-expected-leaks` and `instrument-tests!`.
+
+  To enable it, pass `-Dio.netty.customResourceLeakDetector=aleph.resource_leak_detector` to the
+  JVM. You will also have to pass `-Dio.netty.leakDetection.level=PARANOID`.
+
+  Most reliable results are achieved by also passing the following options:
+  `-Dio.netty.leakDetection.targetRecords=1`
+  `-Dio.netty.allocator.type=unpooled`
+
+  All of the above is also provided by the `:leak-detection` Leiningen profile.
+
+  NOTE: Currently only improves reliability for detecting leaked ByteBufs. Other types of leaked
+  resources will still be detected but not with the same reliability. Search Netty's codebase for
+  invocations of `newResourceLeakDetector` to see which other candidate resources types there are."
+  (:gen-class
+   :extends io.netty.util.ResourceLeakDetector)
+  (:require
+   [clojure.test :as test]
+   [clojure.tools.logging :as log])
+  (:import
+   (io.netty.buffer AbstractByteBufAllocator)
+   (io.netty.util
+    ResourceLeakDetector
+    ResourceLeakDetector$Level
+    ResourceLeakDetectorFactory)))
+
+(defn enabled?
+  "Checks whether the resource leak detector is enabled.
+
+  See `aleph.resource-leak-detector` docstring on how to enable it."
+  []
+  (= "aleph.resource_leak_detector"
+     (System/getProperty "io.netty.customResourceLeakDetector")))
+
+(def active-resource-leak-detector-class
+  (delay
+    (class (.newResourceLeakDetector (ResourceLeakDetectorFactory/instance) String))))
+
+(def active?
+  (delay
+    (= (Class/forName "aleph.resource_leak_detector")
+       @active-resource-leak-detector-class)))
+
+(defn ensure-consistent-config! []
+  (when-not @active?
+    (if (enabled?)
+      (throw (RuntimeException.
+              (str "`aleph.resource_leak_detector` is enabled but the active resource leak detector is "
+                   "`" (.getName ^Class @active-resource-leak-detector-class)"`. This indicates that "
+                   "`io.netty.util.ResourceLeakDetectorFactory` ran into an initialization error. "
+                   "Enable Netty debug logging to diagnose the cause.")))
+      (throw (RuntimeException.
+              (str "Attempted to use `aleph.resource-leak-detector` API but it is not enabled. "
+                   "Pass `-Dio.netty.customResourceLeakDetector=aleph.resource_leak_detector` to enable.")))))
+  (when-not (= ResourceLeakDetector$Level/PARANOID (ResourceLeakDetector/getLevel))
+    (throw (RuntimeException.
+            (str "`aleph.resource_leak_detector` requires `-Dio.netty.leakDetection.level=PARANOID`. "
+                 "Current level is `" (ResourceLeakDetector/getLevel) "`.")))))
+
+(def +max-probe-gc-runs+
+  "Maximum number of times the GC will be run to detect a leaked probe."
+  10)
```

Comment:
Style: No need for Common Lisp-style constants (`+some-const+`) since everything in clj is immutable by default.

Reply:
Might want to consider adding `^:const` though, if they're compile-time literals like numbers or strings, since then they can be inlined.

Reply:
Oh yeah, good call, also done :+1: I didn't mark `max-probe-gc-runs` as const so that one can experiment with it on the REPL (as suggested by the corresponding error raised by `await-probe!`).

Reply:
True, done!